mapping_parenting_tech/analysis/prototyping/play_store/get_play_store_apps.py

```diff
@@ -7,141 +7,413 @@
+                keep_going = False
+            else:
+                continuation_token = log_info[app_id]["continuation_token"]
+                first_pass = False
+                more_to_get = True
+                log_dump = {app_id: log_info[app_id]}
+
+        # if we're forcing a download, then at least download the first set of reviews
+        if force_download:
+            first_pass = True
+            keep_going = True
+
+        while (more_to_get or first_pass) and keep_going:
+            review_fetch, continuation_token = get_playstore_app_reviews(
+                target_app_id=app_id, continuation_token=continuation_token
+            )
+
+            # if we just grabbed a full 200 reviews, there are (probably) more to get
+            more_to_get = True if len(review_fetch) == 200 else False
+
+            # process what we've just downloaded:
+            # 1. add the add id to each review
+            # 2. remove the userName and userImage
+            # 3. add to the target file if we don't already have it
+            with open(target_file, "at", newline="") as csv_file:
+                csv_writer = csv.DictWriter(csv_file, fieldnames=field_names)
+                for review in review_fetch:
+                    review.update({"appId": app_id})
+                    del review["userName"]
+                    del review["userImage"]
+                    if review["reviewId"] not in existing_review_ids:
+                        csv_writer.writerow(review)
+                        existing_review_ids.append(review["reviewId"])
+
+            # save a snapshot of where we are - the app_id and continuation_token - so we can resume later
+            with open(log_file, "rb+") as log_file_handle:
+                if first_pass:
+                    log_dump = {
+                        app_id: {
+                            "completed": False,
+                            "continuation_token": continuation_token,
+                            "latest_review_id": review_fetch[0]["reviewId"],
+                            "latest_review_time": review_fetch[0]["at"],
+                            "downloaded": len(review_fetch),
+                        }
+                    }
+                else:
+                    log_dump[app_id].update(
+                        {
+                            "continuation_token": continuation_token,
+                            "downloaded": (
+                                log_dump[app_id]["downloaded"] + len(review_fetch)
+                            ),
+                        }
+                    )
+                log_info.update(log_dump)
+                pickle.dump(log_info, log_file_handle)
+
+            app_review_count += len(review_fetch)
+
+            if not run_quietly:
+                print(
+                    f"Retrieved {app_review_count} new reviews for {app_id}; fetched {log_dump[app_id]['downloaded']} in total)          ",
```

Comment:
@beingkk Spaces are intentional padding - this line prints in the same position (`end='\r'`) so want to avoid characters remaining if the length of the next print is shorter than the last one.